### PR TITLE
Add support for tracking causes of wide pointers

### PR DESCRIPTION
**Pull request by benharsh**
This commit is the first part of a larger change that will depend on knowing which vars caused other vars to become wide pointers. This first part simply adds support for tracking and printing these causes, and should not impact the number of wide pointers in the program.

Other changes:
- simplify logic for PRIM_RETURN
- move debug timer to earlier location

### CODE REVIEW

**Comment by benharsh** [comment]
passing testing for:
- [ ] local
- [x] no-local
- [x] gasnet

File: compiler/passes/insertWideReferences.cpp [hunk 1]
```diff
@@ -198,8 +198,12 @@
   #define DEBUG_PRINTF(...)
 #endif
 
-static void debug(Symbol* sym, const char* format, ...) {
+static void debug(BaseAST* base, const char* format, ...) {
 #ifdef PRINT_WIDE_ANALYSIS
+  Symbol* sym = toSymbol(base);
+  if (!sym) {
+    sym = toSymExpr(base)->var;
+  }
```
Comment (ronawho) [inline compiler/passes/insertWideReferences.cpp]: add an `INT_ASSERT(sym != NULL)`?

File: compiler/passes/insertWideReferences.cpp [hunk 2]
```diff
@@ -198,8 +198,12 @@
   #define DEBUG_PRINTF(...)
 #endif
 
-static void debug(Symbol* sym, const char* format, ...) {
+static void debug(BaseAST* base, const char* format, ...) {
 #ifdef PRINT_WIDE_ANALYSIS
+  Symbol* sym = toSymbol(base);
+  if (!sym) {
```
Comment (ronawho) [inline compiler/passes/insertWideReferences.cpp]: minor, but `!sym` => `sym != NULL` to match the style of the `sym == NULL` below?

File: compiler/passes/insertWideReferences.cpp [hunk 3]
```diff
@@ -959,20 +1047,20 @@ static void propagateVar(Symbol* sym) {
...
-              matchWide(sym, field->var);
+              matchWide(def, field->var);
             }
             else if (rhs->isPrimitive(PRIM_GET_SVEC_MEMBER) || 
                      rhs->isPrimitive(PRIM_GET_SVEC_MEMBER_VALUE)) {
-              widenTupleField(rhs, sym);
+              widenTupleField(rhs, def);
             }
-            else if (rhs->isResolved()) {
+            else if (rhs->isResolved() && isRef(def)) {
```
Comment (ronawho) [inline compiler/passes/insertWideReferences.cpp]: Is the addition of `&& isRef(def)` a bug fix, or just left in accidentally. If it's intentional, can you add a comment to the PR message?

File: compiler/passes/insertWideReferences.cpp [hunk 4]
```diff
@@ -1989,5 +2079,8 @@ insertWideReferences(void) {
   }
   printf("Summary:\n\twide: %d\n\tnarrow: %d\n", wide, narrow);
 #endif
+
+  printCauses(NULL);
```
Comment (ronawho) [inline compiler/passes/insertWideReferences.cpp]: Is this just to quiet unused function warnings?

**Comment by ronawho** [comment]
Looks awesome! Left a few superficial/minor comments inline, but I don't need to review again unless you want me to.

File: compiler/passes/insertWideReferences.cpp (hunk 3 above)
Comment (benharsh) [inline compiler/passes/insertWideReferences.cpp]: It was accidental, and redundant with the 'isRef(sym)' check above it. I'll remove it shortly.

File: compiler/passes/insertWideReferences.cpp (hunk 4 above)
Comment (benharsh) [inline compiler/passes/insertWideReferences.cpp]: Yes. Would you prefer if I wrapped the function(s) in an ifdef?

Comment (ronawho) [inline compiler/passes/insertWideReferences.cpp]: I think an ifdef or a `//quiet unused function warning` above the  call would be nice,  but up to you if you think it's worth adding either

**Pull request merged**